From 9760786f3fd9770e63a66595226c05b226d2eaac Mon Sep 17 00:00:00 2001
From: retro98boy <retro98boy@qq.com>
Date: Sun, 23 Aug 2026 15:40:38 +0800
Subject: [PATCH 12/18] arm64: dts: rockchip: add RK3562 display nodes

Describe the RK3562 display pipeline used by the KICKPI K3B.

Add the display-subsystem node and the single-VP VOP with its VOP MMU, CSU clock binding, IOMMU, power domain, IOC GRF, and VP0 output port. Add the RK3562 MIPI DSI bridge, its ports, the external D-PHY, clocks, resets, GRF reference, and display power-domain wiring.

Keep the display nodes disabled by default so boards can enable the pipeline from their board-level device tree.
---
 arch/arm64/boot/dts/rockchip/rk3562.dtsi | 92 ++++++++++++++++++++++++
 1 file changed, 92 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk3562.dtsi b/arch/arm64/boot/dts/rockchip/rk3562.dtsi
index a96218176..44687f7f1 100644
--- a/arch/arm64/boot/dts/rockchip/rk3562.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3562.dtsi
@@ -43,6 +43,12 @@ xin24m: clock-xin24m {
 		clock-output-names = "xin24m";
 	};
 
+	display_subsystem: display-subsystem {
+		compatible = "rockchip,display-subsystem";
+		ports = <&vop_out>;
+		status = "disabled";
+	};
+
 	csu: csu {
 		compatible = "rockchip,rk3562-csu";
 		rockchip,clock = <CSU_GMAC_ACLK  1>,
@@ -1368,6 +1374,92 @@ gmac0_mtl_tx_setup: tx-queues-config {
 			};
 		};
 
+		vop: vop@ff400000 {
+			compatible = "rockchip,rk3562-vop";
+			reg = <0x0 0xff400000 0x0 0x2000>, <0x0 0xff405000 0x0 0x1000>;
+			reg-names = "vop", "gamma-lut";
+			interrupts = <GIC_SPI 135 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru ACLK_VOP>,
+				 <&cru HCLK_VOP>,
+				 <&cru DCLK_VOP>;
+			clock-names = "aclk",
+				      "hclk",
+				      "dclk_vp0";
+			rockchip,csu = <&csu CSU_VOP_ACLK>;
+			rockchip,csu-names = "aclk";
+			iommus = <&vop_mmu>;
+			power-domains = <&power RK3562_PD_VO>;
+			rockchip,grf = <&ioc_grf>;
+			assigned-clocks = <&cru DCLK_VOP>;
+			assigned-clock-parents = <&cru PLL_VPLL>;
+			status = "disabled";
+
+			vop_out: ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				vp0: port@0 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <0>;
+				};
+			};
+		};
+
+		vop_mmu: iommu@ff407e00 {
+			compatible = "rockchip,rk3562-iommu", "rockchip,rk3568-iommu";
+			reg = <0x0 0xff407e00 0x0 0x100>;
+			interrupts = <GIC_SPI 135 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru ACLK_VOP>, <&cru HCLK_VOP>;
+			clock-names = "aclk", "iface";
+			#iommu-cells = <0>;
+			power-domains = <&power RK3562_PD_VO>;
+			status = "disabled";
+		};
+
+		dsi: dsi@ffb10000 {
+			compatible = "rockchip,rk3562-mipi-dsi", "snps,dw-mipi-dsi";
+			reg = <0x0 0xffb10000 0x0 0x10000>;
+			interrupts = <GIC_SPI 132 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru PCLK_DSITX>;
+			clock-names = "pclk";
+			resets = <&cru SRST_P_DSITX>;
+			reset-names = "apb";
+			phys = <&video_phy>;
+			phy-names = "dphy";
+			power-domains = <&power RK3562_PD_VO>;
+			rockchip,grf = <&sys_grf>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				dsi_in: port@0 {
+					reg = <0>;
+				};
+
+				dsi_out: port@1 {
+					reg = <1>;
+				};
+			};
+		};
+
+		video_phy: phy@ffb20000 {
+			compatible = "rockchip,rk3562-dsi-dphy", "rockchip,rk3568-dsi-dphy";
+			reg = <0x0 0xffb20000 0x0 0x10000>;
+			clocks = <&cru CLK_MIPIDSIPHY_REF>,
+				 <&cru PCLK_DSIPHY>;
+			clock-names = "ref", "pclk";
+			resets = <&cru SRST_P_DSIPHY>;
+			reset-names = "apb";
+			power-domains = <&power RK3562_PD_VO>;
+			#phy-cells = <0>;
+			status = "disabled";
+		};
+
 		saradc1: adc@ffaa0000 {
 			compatible = "rockchip,rk3562-saradc";
 			reg = <0x0 0xffaa0000 0x0 0x100>;
-- 
2.55.0

